### PR TITLE
Remove unnecessary bottle deviations

### PR DESCRIPTION
**Pull request by MikeMcQuaid**
Where they won't fail the build remove deviations depending on specific hardware
or whether bottling/not rather than the macOS version.

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Comment by fxcoudert** [comment]
Regarding the hardware part, I find the logic easier to follow if the test is on the hardware than on the macOS version. For example:
   
`args << "-DUSE_SSE4_INSTRUCTIONS=ON" if MacOS.version >= :mojave`

doesn't read very clearly I think. Do we have to remember what hardware is mandated by which macOS version? (I don't think it's clearly documented by Apple, who has requirements in terms of computer models, not CPU features). What about people running macOS on virtual machines?

**Comment by MikeMcQuaid** [comment]
> Do we have to remember what hardware is mandated by which macOS version? (I don't think it's clearly documented by Apple, who has requirements in terms of computer models, not CPU features).

We build for Nehalem on Mojave because it was the first version to not support Code 2 Duo processors (which we built for before): https://github.com/Homebrew/brew/blob/master/Library/Homebrew/extend/os/mac/hardware.rb

I'm open to a different approach (or DSL: `nehalem_or_newer?`) or similar that makes this effectively a OS check rather than a specific-CPU-features check. Additionally, it seems likely in future Apple will again up the CPU requirements (and we can up our bottling requirements) so I want to prepare for those days too.

> What about people running macOS on virtual machines?

In this case it's more "what about people running macOS on virtual machines that they've hacked around the hardware requirements of" in which case: we don't attempt to support that use case.

File: Formula/gmp.rb [hunk 1]
```diff
@@ -17,7 +17,7 @@ class Gmp < Formula
   def install
     # Enable --with-pic to avoid linking issues with the static library
     args = %W[--prefix=#{prefix} --enable-cxx --with-pic]
-    args << "--build=core2-apple-darwin#{`uname -r`.to_i}" if build.bottle?
```
Comment (fxcoudert) [inline Formula/gmp.rb]: We can replace `core2` by `nehalem` on Mojave, if we want: it is accepted by the build, and will use specific asm.

File: Formula/ghc@8.2.rb [hunk 2]
```diff
@@ -73,6 +73,8 @@ def install
 
     # MPN_PATH: The lowest common denominator asm paths that work on Darwin,
     # corresponding to Yonah and Merom. Obviates --disable-assembly.
+    # TODO: adjust this based on the fact Mojave only supports Nehalem and
+    # newer processors.
     ENV["MPN_PATH"] = "x86_64/fastsse x86_64/core2 x86_64 generic" if build.bottle?
```
Comment (fxcoudert) [inline Formula/ghc@8.2.rb]: I'd suggest we align this on the `gmp` formula itself: pass `--build=core2-apple-darwin#{`uname -r`.to_i}` (and possibly `nehalem` on Mojave) to the `configure` script. (This in turn sets the `MPN_PATH` to what GMP itself thinks is more suitable for a given processor.)

File: Formula/ghc.rb [hunk 3]
```diff
@@ -86,6 +86,8 @@ def install
 
     # MPN_PATH: The lowest common denominator asm paths that work on Darwin,
     # corresponding to Yonah and Merom. Obviates --disable-assembly.
+    # TODO: adjust this based on the fact Mojave only supports Nehalem and
+    # newer processors.
```
Comment (fxcoudert) [inline Formula/ghc.rb]: I'd suggest we align this on the `gmp` formula itself: pass `--build=core2-apple-darwin#{`uname -r`.to_i}` (and possibly `nehalem` on Mojave) to the `configure` script. (This in turn sets the `MPN_PATH` to what GMP itself thinks is more suitable for a given processor.)

File: Formula/gmp.rb (hunk 1 above)
Comment (MikeMcQuaid) [reply Formula/gmp.rb]: @fxcoudert Good idea. Will hold off until we figure out some sort of better DSL or similar.

**Comment by MikeMcQuaid** [comment]
@fxcoudert Another option is that we use `Hardware.oldest_cpu` which allows you to then check if it's `core2` or `nehalem` or whatever?

**Comment by fxcoudert** [comment]
> Another option is that we use `Hardware.oldest_cpu`

I don't think we can do  `Hardware.oldest_cpu.sse4?` or `Hardware.oldest_cpu.features.include? :vmx`. That would be the way to go, it seems.

**Comment by MikeMcQuaid** [comment]
> I don't think we can do `Hardware.oldest_cpu.sse4?` or `Hardware.oldest_cpu.features.include? :vmx`. That would be the way to go, it seems.

We can't but that could be something that's added. How about `MacOS.oldest_cpu`? Would you mind if this is merged as-is and that's added later? "Yes I mind" is a reasonable response.

**Comment by fxcoudert** [comment]
> Would you mind if this is merged as-is and that's added later?

Not really. But (unrelated) there are some weird test failures, for now.

**Comment by MikeMcQuaid** [comment]
@fxcoudert Stripping out the arch/CPU specific stuff from here for now.

**Comment by MikeMcQuaid** [comment]
Test failures seem unrelated to changes. Will address in follow-up PRs.

**Pull request merged**